src/main/java/com/google/javascript/clutz/DeclarationGenerator.java

```diff
@@ -2385,23 +2385,33 @@ private void visitRecordType(ObjectType type) {
       emit("{");
       Iterator<String> it = getEmittablePropertyNames(type).iterator();
       while (it.hasNext()) {
+
         String propName = it.next();
+        JSType propType = type.getPropertyType(propName);
+
+        if (propType == type) {
```

Comment:
Couldn't you easily have a `Set` of seen objects that acts as a stack, i.e. that you add something into as you're visiting, then remove as you leave?

Reply:
sure, done.